File: Python/N-Queens II.py
"""
Follow up for N-Queens problem.

Now, instead outputting board configurations, return the total number of distinct solutions.
"""
class Solution:
    # @param n, an integer
    # @return an integer
    def totalNQueens(self, n):
        self.column = [False] * n
        self.main_diag = [False] * 2 * n
        self.anti_diag = [False] * 2 * n
        return self.totalNQueensRe(0,[],n)
        
    def totalNQueensRe(self,row,solution,n):
        if row == n:
            return 1
        result = 0
        for i in xrange(n):
            if self.column[i] is False and self.main_diag[row+i] is False and self.anti_diag[row-i+n] is False:
                self.column[i] = self.main_diag[row+i] = self.anti_diag[row-i+n] = True
                result += self.totalNQueensRe(row+1,solution+[i],n)
                self.column[i] = self.main_diag[row+i] = self.anti_diag[row-i+n] = False
        return result

File: Java/N-Queens II.java
/*
Follow up for N-Queens problem.

Now, instead outputting board configurations, return the total number of distinct solutions.
*/

public class Solution {
    private boolean[] main_diag;
    private boolean[] anti_diag;
    /*whether each column is taken*/
    private boolean[] columns;
    private int result = 0;
    
    public int totalNQueens(int n) {
        main_diag = new boolean[2 * n];
        anti_diag = new boolean[2 * n];
        columns = new boolean[n];
        /* the column number in each row */
        int[] used = new int[n];
        
        dfs(used, 0);
        
        return result;
    }
    
    private void dfs(int[]used, int row) {
        int N = used.length;
        
        if(row == N) {
            result ++;
            return;
        }
        
        /*test each columns*/
        for(int j = 0 ; j < N ; j ++) {
            boolean taken = columns[j] || main_diag[row + j] || anti_diag[row - j + N];
            if(taken) continue;
            
            used[row] = j;
            columns[j] = main_diag[row + j] = anti_diag[row - j + N] = true;
            dfs(used, row + 1);
            columns[j] = main_diag[row + j] = anti_diag[row - j + N] = false;
        }
    }
}
